internal/pkg/scaffold/olm-catalog/package_manifest.go

```diff
@@ -0,0 +1,163 @@
+			lowerProjName+PackageManifestPrefix)
+	}
+	return s.Input, nil
+}
+
+var _ scaffold.CustomRenderer = &PackageManifest{}
+
+// SetFS is a no-op to implement CustomRenderer.
+func (s *PackageManifest) SetFS(_ afero.Fs) {}
+
+// CustomRender either reads an existing package manifest or creates a new
+// manifest and modifies it based on values set in s.
+func (s *PackageManifest) CustomRender() ([]byte, error) {
+	i, err := s.GetInput()
+	if err != nil {
+		return nil, err
+	}
+	path := filepath.Join(s.AbsProjectPath, i.Path)
+
+	pm := &olmregistry.PackageManifest{}
+	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
+		pm = s.newPackageManifest()
+	} else if err == nil {
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, errors.Wrapf(err, "read package manifest %s", path)
+		}
+		if len(b) > 0 {
+			if err = yaml.Unmarshal(b, pm); err != nil {
+				return nil, errors.Wrapf(err, "unmarshal package manifest %s", path)
+			}
+		} else {
+			// File exists but is empty.
+			pm = s.newPackageManifest()
+		}
+	} else {
+		return nil, errors.Wrapf(err, "package manifest %s", path)
+	}
+
+	if err = s.setChannels(pm); err != nil {
+		return nil, err
+	}
+
+	sort.Slice(pm.Channels, func(i int, j int) bool {
+		return pm.Channels[i].Name < pm.Channels[j].Name
+	})
+
+	return yaml.Marshal(pm)
+}
+
+func (s *PackageManifest) newPackageManifest() *olmregistry.PackageManifest {
+	// Take the current CSV version to be the "alpha" channel, as an operator
+	// should only be designated anything more stable than "alpha" by a human.
+	channel := "alpha"
+	if s.Channel != "" {
+		channel = s.Channel
+	}
+	pm := &olmregistry.PackageManifest{
+		PackageName: s.ProjectName,
+		Channels: []olmregistry.PackageChannel{
+			{Name: channel, CurrentCSVName: getCSVName(s.ProjectName, s.CSVVersion)},
+		},
+		DefaultChannelName: channel,
+	}
+	return pm
+}
+
+// setChannels checks for duplicate channels in pm and sets the default
+// channel if possible.
+func (s *PackageManifest) setChannels(pm *olmregistry.PackageManifest) error {
+	channelMap := map[string]string{}
+	for _, c := range pm.Channels {
+		if _, ok := channelMap[c.Name]; ok {
+			return fmt.Errorf(`duplicate package manifest channel name "%s"; channel names must be unique`, c.Name)
+		}
+		channelMap[c.Name] = c.CurrentCSVName
+	}
+	if s.Channel != "" {
+		channelMap[s.Channel] = getCSVName(s.ProjectName, s.CSVVersion)
+	}
+	channels := []olmregistry.PackageChannel{}
+	for n, cn := range channelMap {
+		channels = append(channels, olmregistry.PackageChannel{
+			Name:           n,
+			CurrentCSVName: cn,
+		})
+	}
+	pm.Channels = channels
+
+	// Use s.Channel as the default channel if caller has specified it as the
+	// default. If no default channel was specified and there is only one
+	// channel, use that channel as the default.
+	if s.ChannelIsDefault && s.Channel != "" {
+		pm.DefaultChannelName = s.Channel
+	} else if len(pm.Channels) == 1 {
```

Comment:
If `pm.DefaultChannelName != "" && pm.DefaultChannelName != pm.Channels[0].Name`, we would silently overwrite the default channel.

Should we only set the default channel if `len(pm.Channels) == 1 && pm.DefaultChannelName == ""`?

Reply:
We should overwrite the default channel in this case, since the channel in `pm.DefaultChannelName` would only change if it didn't exist in `pm.Channels`, which is an error.

Reply:
> ... which is an error.

Yeah, that's what I was getting at. In the case where there are two channels listed in the `channels` list, where neither of them match the default channel, we don't touch `defaultChannel` and we log a warning that `defaultChannel` does not exist in the `channels` list. IMO, we should do the same thing when there is only one channel.

For example, what if they actual intend for the value of  `defaultChannel` to be in the `channels` list, but accidentally typo'd or changed the `channel` name to something else? We would propagate that typo into `defaultChannel`.

Another related corner case is if one runs with `--csv-channel=stable` with a file containing:
```yaml
defaultChannel: alpha
packageName: packagemanifest-operator
```
We end up overwriting `defaultChannel` the same way:
```yaml
channels:
- currentCSV: packagemanifest-operator.v0.1.1
  name: stable
defaultChannel: stable
packageName: packagemanifest-operator
```

Maybe the solution to most of these corner cases is to validate the existing file before applying changes. And if validation fails, error out with a message that tells the user how to fix the issue?

Reply:
I like the pre-validation. I'll add that check and log a warning on an empty or non-existent default channel.